File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-com-5-4/Lag.cpp
// Lag.cpp 
#include "Lag.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

Lag::Lag(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //lagDurationNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* lagDurationNode = xmlNode->FirstChildElement("lagDuration");

   if(lagDurationNode){lagDurationIsNull_ = false;}
   else{lagDurationIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- lagDurationNode , address : " << lagDurationNode << std::endl;
   #endif
   if(lagDurationNode)
   {
      if(lagDurationNode->Attribute("href") || lagDurationNode->Attribute("id"))
      {
          if(lagDurationNode->Attribute("id"))
          {
             lagDurationIDRef_ = lagDurationNode->Attribute("id");
             lagDuration_ = boost::shared_ptr<Period>(new Period(lagDurationNode));
             lagDuration_->setID(lagDurationIDRef_);
             IDManager::instance().setID(lagDurationIDRef_,lagDuration_);
          }
          else if(lagDurationNode->Attribute("href")) { lagDurationIDRef_ = lagDurationNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { lagDuration_ = boost::shared_ptr<Period>(new Period(lagDurationNode));}
   }

   //firstObservationDateOffsetNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* firstObservationDateOffsetNode = xmlNode->FirstChildElement("firstObservationDateOffset");

   if(firstObservationDateOffsetNode){firstObservationDateOffsetIsNull_ = false;}
   else{firstObservationDateOffsetIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- firstObservationDateOffsetNode , address : " << firstObservationDateOffsetNode << std::endl;
   #endif
   if(firstObservationDateOffsetNode)
   {
      if(firstObservationDateOffsetNode->Attribute("href") || firstObservationDateOffsetNode->Attribute("id"))
      {
          if(firstObservationDateOffsetNode->Attribute("id"))
          {
             firstObservationDateOffsetIDRef_ = firstObservationDateOffsetNode->Attribute("id");
             firstObservationDateOffset_ = boost::shared_ptr<Period>(new Period(firstObservationDateOffsetNode));
             firstObservationDateOffset_->setID(firstObservationDateOffsetIDRef_);
             IDManager::instance().setID(firstObservationDateOffsetIDRef_,firstObservationDateOffset_);
          }
          else if(firstObservationDateOffsetNode->Attribute("href")) { firstObservationDateOffsetIDRef_ = firstObservationDateOffsetNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { firstObservationDateOffset_ = boost::shared_ptr<Period>(new Period(firstObservationDateOffsetNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<Period> Lag::getLagDuration()
{
   if(!this->lagDurationIsNull_){
        if(lagDurationIDRef_ != ""){
             return boost::shared_static_cast<Period>(IDManager::instance().getID(lagDurationIDRef_));
        }else{
             return this->lagDuration_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<Period>();
   }
}
boost::shared_ptr<Period> Lag::getFirstObservationDateOffset()
{
   if(!this->firstObservationDateOffsetIsNull_){
        if(firstObservationDateOffsetIDRef_ != ""){
             return boost::shared_static_cast<Period>(IDManager::instance().getID(firstObservationDateOffsetIDRef_));
        }else{
             return this->firstObservationDateOffset_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<Period>();
   }
}
}

File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-com-5-4/Lag.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class Lag : ISerialized
    {
        public Lag(XmlNode xmlNode)
        {
            XmlNode lagDurationNode = xmlNode.SelectSingleNode("lagDuration");
            
            if (lagDurationNode != null)
            {
                if (lagDurationNode.Attributes["href"] != null || lagDurationNode.Attributes["id"] != null) 
                {
                    if (lagDurationNode.Attributes["id"] != null) 
                    {
                        lagDurationIDRef_ = lagDurationNode.Attributes["id"].Value;
                        Period ob = new Period(lagDurationNode);
                        IDManager.SetID(lagDurationIDRef_, ob);
                    }
                    else if (lagDurationNode.Attributes["href"] != null)
                    {
                        lagDurationIDRef_ = lagDurationNode.Attributes["href"].Value;
                    }
                    else
                    {
                        lagDuration_ = new Period(lagDurationNode);
                    }
                }
                else
                {
                    lagDuration_ = new Period(lagDurationNode);
                }
            }
            
        
            XmlNode firstObservationDateOffsetNode = xmlNode.SelectSingleNode("firstObservationDateOffset");
            
            if (firstObservationDateOffsetNode != null)
            {
                if (firstObservationDateOffsetNode.Attributes["href"] != null || firstObservationDateOffsetNode.Attributes["id"] != null) 
                {
                    if (firstObservationDateOffsetNode.Attributes["id"] != null) 
                    {
                        firstObservationDateOffsetIDRef_ = firstObservationDateOffsetNode.Attributes["id"].Value;
                        Period ob = new Period(firstObservationDateOffsetNode);
                        IDManager.SetID(firstObservationDateOffsetIDRef_, ob);
                    }
                    else if (firstObservationDateOffsetNode.Attributes["href"] != null)
                    {
                        firstObservationDateOffsetIDRef_ = firstObservationDateOffsetNode.Attributes["href"].Value;
                    }
                    else
                    {
                        firstObservationDateOffset_ = new Period(firstObservationDateOffsetNode);
                    }
                }
                else
                {
                    firstObservationDateOffset_ = new Period(firstObservationDateOffsetNode);
                }
            }
            
        
        }
        
    
        #region LagDuration_
        private Period lagDuration_;
        public Period LagDuration_
        {
            get
            {
                if (this.lagDuration_ != null)
                {
                    return this.lagDuration_; 
                }
                else if (this.lagDurationIDRef_ != null)
                {
                    lagDuration_ = IDManager.getID(lagDurationIDRef_) as Period;
                    return this.lagDuration_; 
                }
                else
                {
                      return this.lagDuration_; 
                }
            }
            set
            {
                if (this.lagDuration_ != value)
                {
                    this.lagDuration_ = value;
                }
            }
        }
        #endregion
        
        public string lagDurationIDRef_ { get; set; }
        #region FirstObservationDateOffset_
        private Period firstObservationDateOffset_;
        public Period FirstObservationDateOffset_
        {
            get
            {
                if (this.firstObservationDateOffset_ != null)
                {
                    return this.firstObservationDateOffset_; 
                }
                else if (this.firstObservationDateOffsetIDRef_ != null)
                {
                    firstObservationDateOffset_ = IDManager.getID(firstObservationDateOffsetIDRef_) as Period;
                    return this.firstObservationDateOffset_; 
                }
                else
                {
                      return this.firstObservationDateOffset_; 
                }
            }
            set
            {
                if (this.firstObservationDateOffset_ != value)
                {
                    this.firstObservationDateOffset_ = value;
                }
            }
        }
        #endregion
        
        public string firstObservationDateOffsetIDRef_ { get; set; }
        
    
        
    
    }
    
}

